### PR TITLE
drivers/pwm: Add driver to support RPM2PWM IP

### PR DESCRIPTION
**Pull request by jjdalynh**
The MEC172x series of Embedded Controllers from Microchip includes a block to control up to 2 fans via a closed loop tach rpm to pwm mechanism.  This (proposed) driver adds the basic functionality to support fans using this hardware.

### CODE REVIEW

**Comment by MaureenHelm** [comment]
@jvasanth1 @albertofloyd can you take a look please?

**Comment by jjdalynh** [comment]
this driver works, but it could probably be expanded some to support different fan/platform configuration options from DTS.  currently it's pretty fixed with the RPM control configuration, and only taking the number of fan edges from DTS for configuration.  there are options for spin up config, update period, fan step values, etc.  These don't have bits defined in the PWM dt spec, and there's only 8 bits reserved for hardware-specific configuration, so care would have to be taken if some configuration flags need defining.

File: drivers/pwm/rpm2pwm_mchp_xec.c [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+	ARG_UNUSED(chan);
+
+	const struct rpm2pwm_tach_xec_config * const cfg = dev->config;
+	struct device * const parent = cfg->parent;
+	const struct rpm2pwm_xec_config * const parent_cfg = parent->config;
+	struct rpm2pwm_regs * const regs = parent_cfg->regs;
+
+	struct rpm2pwm_tach_xec_data * const data = dev->data;
+
+	data->count = 3932160 / (regs->TACH >> 3);
```
Comment (albertofloyd) [inline drivers/pwm/rpm2pwm_mchp_xec.c:74]: Magic numbers, add macros or device tree entries as applicable

File: drivers/pwm/rpm2pwm_mchp_xec.c [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+	uint8_t spinup;
+	uint16_t config;
+
+	spinup = regs->SPINUP;
+
+	/*
+	 * TODO: Figure out DTS flag options that are appropriate:
+	 * if (flags & RPM2PWM_XEC_FLAG_SPIN_NOKICK)
+	 */
+	spinup |= (1 << 5);
```
Comment (albertofloyd) [inline drivers/pwm/rpm2pwm_mchp_xec.c:126]: Magic numbers, need to use dts flag or macro applicable

File: drivers/pwm/rpm2pwm_mchp_xec.c [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+
+	if (channel > 0) {
+		return -EIO;
+	}
+
+	if (cycles) {
+		/* User does not have to know about lowest clock,
+		 * the driver will select the most relevant one.
+		 */
+		*cycles = 32768;
```
Comment (albertofloyd) [inline drivers/pwm/rpm2pwm_mchp_xec.c:164]: magic numbers, there is an API that returns cycles per second if that's the intention here.  sys_clock_hw_cycles_per_sec

File: dts/bindings/pwm/microchip,xec-rpm2pwm.yaml [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  The fan controllers use an input tachometer reading and a target RPM to control the PWM to the
+  fans.  There are multiple fan type configurations to support fans with different numbers of poles
+  in order to read the RPM correctly, as well as options to control the fan ramp up/down and
+  startup spin control.
+
+  This driver is an initial baseline requesting review and comment on the design and feature-set.
+
+  Example DTS board configuration:
+
+  &rpmfan0 {
```
[the remark this answers is not in the document]
Comment (jjdalynh) [reply dts/bindings/pwm/microchip,xec-rpm2pwm.yaml:18]: does this need to go under samples/drivers or tests/drivers ?

[the remark this answers is not in the document]
Comment (albertofloyd) [reply dts/bindings/pwm/microchip,xec-rpm2pwm.yaml:18]: Either way

**Pull request closed without merging**